rolling-shutter/collator/batch/batchhandler.go

```diff
@@ -0,0 +1,503 @@
+		return 0, err
+	}
+	return shdb.DecodeUint64(epochID), nil
+}
+
+func NewBatchHandler(cfg config.Config, dbpool *pgxpool.Pool) (*BatchHandler, error) {
+	ctx := context.Background()
+
+	l2Client, err := rpc.Dial(cfg.SequencerURL)
+	if err != nil {
+		return nil, err
+	}
+	l2EthClient := ethclient.NewClient(l2Client)
+	chainID, err := l2EthClient.ChainID(ctx)
+	if err != nil {
+		return nil, err
+	}
+	signer := txtypes.LatestSignerForChainID(chainID)
+
+	return &BatchHandler{
+		l2Client:    l2Client,
+		l2EthClient: l2EthClient,
+		config:      cfg,
+		signer:      signer,
+		txpool:      NewTransactionPool(signer),
+		dbpool:      dbpool,
+	}, nil
+}
+
+// BatchHandler is a threadsafe handler to process the following actions
+// - validate and enqueue incoming tx's either to the currently pending batch
+//    or for later validation in the transaction pool
+// - handle incoming decryption keys and send the currently pending batch-transaction
+//    to the sequencer
+// - start next epoch and initiate the broadcasting of the decryption-trigger to
+//    the keypers
+type BatchHandler struct {
+	mux      sync.Mutex
+	l2Client *rpc.Client
+	// l2EthClient is the RPC l2Client wrapped by ethclient.Client
+	l2EthClient *ethclient.Client
+	config      config.Config
+	txpool      *TransactionPool
+	signer      txtypes.Signer
+	dbpool      *pgxpool.Pool
+	latestBatch *Batch
+}
+
+func (bh *BatchHandler) Signer() txtypes.Signer {
+	return bh.signer
+}
+
+// LatestEpochID returns the epoch-id associated to
+// the current batch that is yet to be submitted to the sequencer.
+func (bh *BatchHandler) LatestEpochID() uint64 {
+	return bh.latestBatch.EpochID()
+}
+
+func (bh *BatchHandler) privateKey() *ecdsa.PrivateKey {
+	return bh.config.EthereumKey
+}
+
+// EnqueueTx handles the potential addition of a user's transaction
+// to the latest local batch or in case of future transactions to the transaction pool.
+// Future transactions can't be queued if they have a batch-index that is too
+// far in the future - the allowed difference is set by the `TransactionAcceptanceInterval` (const).
+// Transactions that are queued in the transaction pool can't be validated yet,
+// because the chain-state (balance, nonce) at the future batch is yet to be known.
+// Thus, even successfully queued transactions can't be guaranteed to be considered
+// later on for inclusion in the corresponding batch.
+// Due to the limitation that a user HAS to encrypt a transaction to a very specific
+// batch, a global priority-queue like transaction-pool like on Layer1 ethereum
+// is not possible. Users have to re-submit a transaction that did not make it
+// to the specific batch explicitly.
+func (bh *BatchHandler) EnqueueTx(ctx context.Context, txBytes []byte) error {
+	var tx txtypes.Transaction
+	err := tx.UnmarshalBinary(txBytes)
+	if err != nil {
+		return errors.New("can't decode transaction bytes")
+	}
+
+	if tx.BatchIndex() > math.MaxUint32 {
```

Comment:
Why is that? The integration (sequencer) is not working anyways currently, or is it?
The problem here is that normal transactions are not working yet (see PR comment) because we don't distinguish the tx-types and rely on shutter-tx exclusive fields.

Reply:
or did intend to write "that's **not** a `ShutterTX` ?